### DISCUSSION TITLE
Is the `path` integrator based on autodiff?

### DISCUSSION THREAD

**Discussion by gerwang** [mitsuba-renderer/mitsuba3#691] [Q&A] [answered]
Hi @njroussel

Could you please tell me whether the `path` integrator is based on autodiff or radiative backpropagation during `dr.backward()`? Are the forward and backward processes statistically independent? What about the `direct_reparam` integrator? Thank you!

**Comment by njroussel** [comment]
Hi @gerwang 

Let me point you to a few different sources so you can check this for yourself too:

* The `mi.render` ([source](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/util.py#L406)) function creates a `CustomOp` which defines how `dr.forward()` and `dr.backward()` should behave on this operation.
* The `CustomOp` for the rendering operation is defined [here](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/util.py#L352-L404): `RenderOp`. Basically, it calls `integrator.render_backward()` for `dr.backward()`. Reminder: `forward` here means forward-mode differentiation and not the typical "forward" process which we call "primal" to disambiguate.
* For non-`RBIntegrator` types (not radiative backropagation style integrators) the `integrator.render_backward()` method is defined [here](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/ad/integrators/common.py#L1065-L1144) and basically just used automatic differentiation as you would expect.
* For `RBIntegrator` integrators, the `integrator.render_backward()` implementation is located [here](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/ad/integrators/common.py#L784-L971). As you can see, [here](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/ad/integrators/common.py#L924-L948) we do the two rendering passes: primal and backward mode.

Now for your questions:
1) `path` is not a `RBIntegrator` - it uses automatic differentiation.
2) When creating the `RenderOp`, both processes use different sampler seeds ([source](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/util.py#L511-L513)).
3) The `direct_reparam` is in a [bit of an odd spot because of its short path length](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/ad/integrators/direct_reparam.py#L46-L51). We handle it differently.

**Comment by gerwang** [comment]
There remains one thing not intuitive to me: If automatic differentiation is used, is there anything involved with randomness? Does it make a difference whether we set different seeds in primal and backward modes?

**Reply by njroussel** [reply, marked as answer]
Even with automatic differentiation, there are still practically two renderings:
* The first is typically used to compute the loss (this is the `eval()` path in the `RenderOp`).
* The second one happens once you call `dr.backward(loss)` (it will trigger the `backward()` path in the `RenderOp`). This one will use automatic differentiation on its output but it must still be de-correlated from the primal.

Does that make sense?
